### ISSUE TITLE
"expected" and "actual" are still swapped?

### ISSUE THREAD

**Issue by gelisam** [gelisam/klister#128]
I remember that our error messages used to say "expected ACTUAL but found EXPECTED". We fixed that, but the error message is still backwards:

```
-- Type mismatch at unknown location. Expected Syntax but got (List Integer)
(example (the (List Integer) 'foo))
```

If I had written the type-checker myself, I would definitely have used unification so that I wouldn't have to think about when to check and when to synthesize. And then I would expect to see bugs like the above, because there is no way to know which side of a unification is the expected type and which side is the actual type. But since @david-christiansen wrote the type-checker, I'm pretty sure he used a bidirectional algorithm, in which we do know which side is expected and which side is actual. So what's going on?

Also, why is the location unknown? The very similar `(example (+ 'foo 1))` does give a location.

**Comment by david-christiansen** [comment]
The type checker is not written bidirectionally. It's a common variation on the usual Hindley-Milner style.

In textbook H-M, the type checker is really a type inferencer. That is, you give it a context and a term, and it monadically returns a type. A downside of this approach is that it creates tons of metavariables, because if something's type is not restricted, it has to create a meta to say so. Then, the calling context unifies the returned thing with the type that it had to have anyway. The version used in Klister is one where the type checker  takes a context, a term, and a type, and either succeeds or fails monadically. The behavior of the inferencer can be recovered by passing this version a fresh meta, then checking what it's been solved to afterwards. But now a calling context can pass a type inwards, which saves on fresh metas but also allows some more control over error messages.

The bug here is just two arguments flipped to unify somewhere.

**Comment by david-christiansen** [comment]
As far as unknown location goes, the type checker presently keeps track of source locations in a table indexed by SplitCorePtr. Some operator somewhere isn't filling in the table for itself.

**Comment by david-christiansen** [comment]
Fixed in #129.
